spotpass: Add task finder scripts

MisterSheeple opened the pull request:
This is a script that automatically checks all present app IDs for a specified task name. This is useful for common names such as `FGONLYT` where many app IDs use it. This is essentially a mish-mash of parts of the `read-boss-db-3ds` and `scrape-3ds` scripts. It pings NPFL to check if the task exists under a certain app ID. If it's considered valid for that app ID, then it adds that task to the json. If invalid, it ignores and moves on.

**TODO**:

- [x] Everything Wii U (to be done once all of the below todos are finished)
- [x] Remove dependency on database and do things similarly to how `3ds-tasks-to-tid` does it
- [x] Write tasks one by one rather than all at once
- [x] Make it skip the NPFL process for apps that are already known to have the searched task.
- [x] Make it check existing tasks regardless of case

It's kinda sloppy right now, so feedback is greatly appreciated.

MisterSheeple committed:
spotpass: Add 3DS task finder script

MisterSheeple committed:
spotpass: Allow build-database to be called as a module

MisterSheeple committed:
spotpass: Fix a few things in 3ds-task-finder

MisterSheeple committed:
spotpass: Remove inbuilt database building from 3DS task finder

MisterSheeple committed:
spotpass: Revert previous changes on build-database

MisterSheeple committed:
spotpass: Write tasks one by one rather than all at once for 3DS task finder

MisterSheeple committed:
spotpass: Remove trailing whitespace in build-database

MisterSheeple committed:
fix previous commit

MisterSheeple committed:
task-finder: Fix else statement

jonbarrow commented on spotpass/3ds-task-finder.js:
```diff
@@ -0,0 +1,59 @@
+
+	if (!response.headers['content-type'] || !response.headers['content-type'].startsWith('text/plain')) {
+		return;
+	} else {
+        	for (const app of apps) {
+        		if (app.app_id === task.app_id) {
+        			if (!app.tasks.includes(TASK_SEARCH)) {
+        				app.tasks.push(TASK_SEARCH);
+						fs.writeJSONSync('./ctr-boss-apps.json', apps, {
+							spaces: '\t'
+						});
+        			}
+        		}
+        	}
+	}
```
Since there's an early return, you can remove this nested else

jonbarrow commented on spotpass/3ds-task-finder.js:
```diff
@@ -0,0 +1,59 @@
+const apps = require('./ctr-boss-apps.json');
+
+const NPFL_URL_BASE = 'https://npfl.c.app.nintendowifi.net/p01/filelist';
+const TASK_SEARCH = 'FGONLYT'; // * For this example, I'm using the common FGONLYT task name, but there are other common task names that can be tested.
```
I'm curious what other common task names you've noticed? `FGONLYT` is the only one that comes to mind for me, which I assume stands for "***F***ore***g***round ***only*** ***t***ask"

MisterSheeple replied:
`news` is another one I noticed that's extremely common.

jonbarrow replied:
You can probably find a few more by just checking the frequency at each task name shows up

jonbarrow replied:
Update: Just wrote a quick script to check this. These are the tasks which show up more than 10 times (checked both Wii U and 3DS tasks)

- `FGONLYT`
- `news`
- `data`
- `TASK00`
- `0000001`

MisterSheeple replied:
Nice, thanks. Should be implemented now.

MisterSheeple committed:
task-finder: Remove else

MisterSheeple committed:
task-finder: Use multiple tasks at once

MisterSheeple commented:
Do I need to input region/lang in order to determine if the task works or can I just remove that part?

jonbarrow commented:
Do you mean in general, or the query params?

MisterSheeple commented:
The query params. Essentially I'm wondering if some tasks specifically don't show up in certain regions/langs or if it's just the files that do that.

jonbarrow commented:
Whether or not the query params are needed depends on 2 things:

1. The endpoint being hit
2. The task

Some tasks *are* region locked. And there's no real way to know which ones are without trying the locale combinations. Additionally some tasks seem to have a "default" response they'll use if an invalid locale is used, while some just 403. Also a task may only be language locked, but not country locked. It's super context dependant and honestly a mess

If a task is region locked, then you only need the query param(s) when using NPFL. NPDL can also be region locked, but it uses the path to determine this rather than query params and there is 4 known variations of this path format (again with no real way to know which is used besides just trying them all)

MisterSheeple committed:
task-finder: Remove row processed

This prevented multiple tasks from being searched. I plan on moving away from using the db in the future but for now this makes it work.

MisterSheeple committed:
task-finder: undo previous commit

MisterSheeple committed:
spotpass: Rewrite 3DS task finder

This should fix most issues and also eliminate the dependency on the database.

MisterSheeple commented:
Please review and let me know if there's any issues. If not, I'll get to work on creating the Wii U equivalent.

MisterSheeple committed:
task-finder: Remove a few things that were used for testing

jonbarrow commented on spotpass/3ds-task-finder.js:
```diff
@@ -0,0 +1,73 @@
+			// * specific combinations. Try every country/language
+			// * combination until a task is found
+			let taskFound = true;
+			for (const country of COUNTRIES) {
+				for (const language of LANGUAGES) {
+					if (await taskExists(app, task, country, language)) {
+						console.log(`Task ${task} found for app id ${app.app_id}`);
+						app.tasks.push(task);
+						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
+							spaces: '\t'
+						});
+
+						taskFound = true;
+						break;
+					}
+				}
+			}
```
`break` in this instance will only break out of the inner loop, not the outer loop. So even when a task is found it will continue looping through any remaining countries, despite not needing to do so.

There is a number of ways in JavaScript to fix this

### 1. Labels
Simply wrapping the outer loop in a [label](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) and using `break` to jump out of it will exit both loops:

```suggestion
			check_locales: for (const country of COUNTRIES) {
				for (const language of LANGUAGES) {
					if (await taskExists(app, task, country, language)) {
						console.log(`Task ${task} found for app id ${app.app_id}`);
						app.tasks.push(task);
						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
							spaces: '\t'
						});

						taskFound = true;
						break check_locales;
					}
				}
			}
```

### 2. Using a standard `for` loop with an extra condition
`for` loops in JavaScript can do more than one thing in all 3 sections. In this case, creating an extra variable in the `initialization` section and doing an extra check in the `condition` section:

```suggestion
			for (let i = 0, deepExit = false; i < COUNTRIES.length && !deepExit; i++) {
				const country = COUNTRIES[i];

				for (const language of LANGUAGES) {
					if (await taskExists(app, task, country, language)) {
						console.log(`Task ${task} found for app id ${app.app_id}`);
						app.tasks.push(task);
						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
							spaces: '\t'
						});

						taskFound = true;
						deepExit = true;
						break;
					}
				}
			}
```

Or reusing the `taskFound` variable to do the same job.

### 3. Wrapping the loop in a function
Wrapping the loop in a function, either self-executing or a standard function declaration, and then using `return` to exit:

```suggestion
			// Self-executing
			(() => {
				for (const country of COUNTRIES) {
					for (const language of LANGUAGES) {
						if (await taskExists(app, task, country, language)) {
							console.log(`Task ${task} found for app id ${app.app_id}`);
							app.tasks.push(task);
							await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
								spaces: '\t'
							});
	
							taskFound = true;
							return;
						}
					}
				}
			})();
```

```suggestion
			// Standard function declaration
			function checkTaskLocales() {
				for (const country of COUNTRIES) {
					for (const language of LANGUAGES) {
						if (await taskExists(app, task, country, language)) {
							console.log(`Task ${task} found for app id ${app.app_id}`);
							app.tasks.push(task);
							await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
								spaces: '\t'
							});
	
							return true;
						}
					}
				}
				
				return false;
			}
```

MisterSheeple replied:
That's weird because I thought I fixed that. I'll have a look.

jonbarrow commented on spotpass/3ds-task-finder.js:
```diff
@@ -0,0 +1,73 @@
+	for (const app of apps) {
+		for (const task of TASKS) {
+			// * Skip the task if the app already has it
+			let titleHasTask = false;
+			for (const appTask of app.tasks) {
+				if (task.toLowerCase() == appTask.toLowerCase()) {
+					titleHasTask = true;
+					break;
+				}
+			}
+
+			if (titleHasTask) {
+				continue;
+			}
```
Not really an issue, just a small suggestion for a possible refactor. You could remove the need for this `titleHasTask` variable, the check with it, and the `continue` statement by using a label on the `TASKS` loop and breaking to it in side the innermost loop.

It's fine as-is, just letting you know the option exists.

MisterSheeple replied:
Eh, it works, and ultimately this script isn't gonna be used beyond 2-3 weeks from now anyway so I don't think it matters much.

jonbarrow reviewed (commented):
Generally fine, only one potential issue

MisterSheeple committed:
task-finder: Fix breakage

MisterSheeple committed:
task-finder: Remove commented out code

jonbarrow commented on spotpass/3ds-task-finder.js:
```diff
@@ -0,0 +1,69 @@
+			// * Most BOSS apps are region-agnostic, but some require
+			// * specific combinations. Try every country/language
+			// * combination until a task is found
+			let taskFound = true;
```
This is unused now

MisterSheeple committed:
spotpass: Add Wii U task finder script

MisterSheeple committed:
task-finder-3ds: Remove unused bit

MisterSheeple committed:
task-finder-wiiu: Remove unused bit

MisterSheeple committed:
task-finder-wiiu: Remove testing bit

MisterSheeple commented:
I think it should be pretty much ready for merge now unless you have any other feedback
